### PR TITLE
Heavy refactoring

### LINKED ISSUE #12: Support Gutenberg (JavaScript I18N)
It would be great if we could automatically create the additional POT file if https://github.com/WordPress/packages/tree/master/packages/babel-plugin-makepot is being used or at least make it so that we can merge the generated POT file with an existing one.

For the latter option, we'd need a way for plugins to tell Traduttore where to find that POT file (config file?) unless we want to scan the whole directory first for existing POT files.

### LINKED ISSUE #37: ZipProvider::get_last_build_time() should not be static
There's no reason for `ZipProvider::get_last_build_time()` to be static.

Everywhere we're using it, we already instantiate a new `ZipProvider` object anyway.

### LINKED ISSUE #40: Use dot as a separator in hook names
Per discussion at https://github.com/wearerequired/traduttore/pull/38#discussion_r198785874

If BC is a concern, we can use `do_action_deprecated` and `do_action` in tandem for a while.

### LINKED ISSUE #47: Make Traduttore provider agnostic
It wouldn't be hard to make Traduttore support more than GitHub, e.g. also GitLab or Bitbucket.

### PR DESCRIPTION
**Pull request by swissspidy**
Fixes #12.
Fixes #37.
Fixes #40.
Fixes #47.

### CODE REVIEW

File: inc/TranslationApiRoute.php [hunk 1]
```diff
@@ -18,17 +18,17 @@
 /**
  * Class used to add a simple translations API.
  *
- * @since 2.0.0
+ * @since 3.0.0
```
Comment (ocean90) [inline inc/TranslationApiRoute.php]: Why the version change?

File: inc/Updater.php [hunk 2]
```diff
@@ -0,0 +1,193 @@
+<?php
+/**
+ * Loader class.
```
Comment (ocean90) [inline inc/Updater.php]: Wrong class name.

File: inc/Updater.php [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+ * @package Required\Traduttore
+ */
+
+namespace Required\Traduttore;
+
+use GP;
+use PO;
+
+/**
+ * LoaderFactory class.
```
Comment (ocean90) [inline inc/Updater.php]: Wrong class name.

File: inc/Updater.php [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+		unlink( $pot_file );
+
+		if ( ! $result ) {
+			return false;
+		}
+
+		$stats = GP::$original->import_for_project( $this->project->get_project(), $translations );
+
+		/**
+		 * Fires after translations have been updated from GitHub.
```
Comment (ocean90) [inline inc/Updater.php]: ~~from GitHub.~~

File: inc/Updater.php [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+		 *
+		 * @since 3.0.0
+		 *
+		 * @param Project $project      The GlotPress project that was updated.
+		 * @param array   $stats        Stats about the number of imported translations.
+		 * @param PO      $translations PO object containing all the translations from the POT file.
+		 */
+		do_action( 'traduttore_updated', $this->project, $stats, $translations );
```
Comment (ocean90) [inline inc/Updater.php]: `traduttore.updated`

**Comment by swissspidy** [comment]
#yolo

**Pull request merged**